File: tools/clod-poc/tools/qa-orchestrator.ts
import { resolve } from "node:path";
import { loadUnifiedRegistry } from "../src/qa/unified/manifest.js";
import { loadQaOrchestration } from "../src/qa/unified/orchestration_manifest.js";
import { runQaBattery } from "../src/qa/unified/battery_runner.js";
import { runQaDeterminism } from "../src/qa/unified/determinism.js";

const REPOSITORY_ROOT = resolve(import.meta.dirname, "../../..");
const DEFAULTS = {
  visual: resolve(REPOSITORY_ROOT, "validation/manifests/visual-regression.yaml"),
  performance: resolve(REPOSITORY_ROOT, "validation/manifests/performance-regression.yaml"),
  legacyMap: resolve(REPOSITORY_ROOT, "validation/manifests/legacy-id-map.yaml"),
  commands: resolve(REPOSITORY_ROOT, "validation/manifests/command-allowlist.yaml"),
  batteries: resolve(REPOSITORY_ROOT, "validation/manifests/batteries.yaml"),
};

interface Args {
  mode: "validate" | "run" | "determinism";
  battery: string;
  output: string;
  target?: "clod-poc" | "bevy";
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  const scenes = loadUnifiedRegistry({ visual: DEFAULTS.visual, performance: DEFAULTS.performance, legacyMap: DEFAULTS.legacyMap });
  const orchestration = loadQaOrchestration({ commands: DEFAULTS.commands, batteries: DEFAULTS.batteries }, scenes);
  console.log(`[qa-orchestrator] validated scenes=${scenes.scenes.length} commands=${orchestration.commands.size} batteries=${orchestration.batteries.size}`);
  if (args.mode === "validate") return;
  const outputDir = resolve(REPOSITORY_ROOT, args.output);
  const report = args.mode === "determinism"
    ? await runQaDeterminism(orchestration, scenes, { repositoryRoot: REPOSITORY_ROOT, outputDir, batteryId: args.battery, target: args.target })
    : await runQaBattery(orchestration, scenes, { repositoryRoot: REPOSITORY_ROOT, outputDir, batteryId: args.battery, target: args.target, runIndex: 1 });
  console.log(`[qa-orchestrator] ${args.mode} status=${report.status}`);
  if (report.status !== "PASS") process.exitCode = 1;
}

function parseArgs(argv: readonly string[]): Args {
  const args: Args = { mode: "validate", battery: "combined-smoke", output: "validation-runs/orchestrated/latest" };
  for (let index = 0; index < argv.length; index++) {
    const arg = argv[index];
    const value = argv[index + 1];
    if (arg === "--mode" && value) {
      if (value !== "validate" && value !== "run" && value !== "determinism") throw new Error(`invalid mode ${value}`);
      args.mode = value;
      index++;
    } else if (arg === "--battery" && value) { args.battery = value; index++; }
    else if (arg === "--output" && value) { args.output = value; index++; }
    else if (arg === "--target" && value) {
      if (value !== "clod-poc" && value !== "bevy") throw new Error(`invalid target ${value}`);
      args.target = value;
      index++;
    } else throw new Error(`unknown or incomplete argument: ${arg}`);
  }
  return args;
}

main().catch((error: unknown) => {
  console.error("[qa-orchestrator] error:", error instanceof Error ? error.message : error);
  process.exit(1);
});

File: src/bin/qa_orchestrator.rs
use std::path::PathBuf;
use std::process::ExitCode;

use clap::{Parser, ValueEnum};
use voxel_builder::diagnostics::qa::unified::determinism::{DeterminismOptions, run_determinism};
use voxel_builder::diagnostics::qa::unified::manifest::load_registry_with_legacy;
use voxel_builder::diagnostics::qa::unified::orchestration::{
    BatteryRunOptions, load_orchestration, run_battery,
};
use voxel_builder::diagnostics::qa::unified::schema::Target;

#[derive(Clone, Copy, Debug, ValueEnum)]
enum Mode { Validate, Run, Determinism }

#[derive(Clone, Copy, Debug, ValueEnum)]
enum TargetArg { ClodPoc, Bevy }

impl From<TargetArg> for Target {
    fn from(value: TargetArg) -> Self {
        match value { TargetArg::ClodPoc => Target::ClodPoc, TargetArg::Bevy => Target::Bevy }
    }
}

#[derive(Parser, Debug)]
#[command(about = "Run unified QA-U5 through QA-U8 orchestration", version)]
struct Args {
    #[arg(long, value_enum, default_value_t = Mode::Validate)]
    mode: Mode,
    #[arg(long, default_value = "combined-smoke")]
    battery: String,
    #[arg(long, value_enum)]
    target: Option<TargetArg>,
    #[arg(long, default_value = "validation-runs/orchestrated/latest")]
    output: PathBuf,
    #[arg(long, default_value = "validation/manifests/visual-regression.yaml")]
    visual: PathBuf,
    #[arg(long, default_value = "validation/manifests/performance-regression.yaml")]
    performance: PathBuf,
    #[arg(long, default_value = "validation/manifests/legacy-id-map.yaml")]
    legacy_map: PathBuf,
    #[arg(long, default_value = "validation/manifests/command-allowlist.yaml")]
    commands: PathBuf,
    #[arg(long, default_value = "validation/manifests/batteries.yaml")]
    batteries: PathBuf,
}

fn main() -> ExitCode {
    match run() {
        Ok(status) if status == "PASS" => ExitCode::SUCCESS,
        Ok(status) => { eprintln!("[qa-orchestrator] status={status}"); ExitCode::from(1) }
        Err(error) => { eprintln!("[qa-orchestrator] error: {error}"); ExitCode::from(1) }
    }
}

fn run() -> Result<String, String> {
    let args = Args::parse();
    let repository_root = std::env::current_dir().map_err(|error| error.to_string())?;
    let scenes = load_registry_with_legacy(&args.visual, &args.performance, Some(&args.legacy_map))
        .map_err(|error| error.to_string())?;
    let orchestration = load_orchestration(&args.commands, &args.batteries, Some(&scenes))
        .map_err(|error| error.to_string())?;
    println!(
        "[qa-orchestrator] validated scenes={} commands={} batteries={}",
        scenes.scenes.len(), orchestration.commands.len(), orchestration.batteries.len()
    );
    match args.mode {
        Mode::Validate => Ok("PASS".to_string()),
        Mode::Run => {
            let report = run_battery(
                &orchestration,
                &scenes,
                &BatteryRunOptions {
                    repository_root,
                    output_dir: args.output,
                    run_index: 1,
                    battery_id: args.battery,
                    target: args.target.map(Into::into),
                },
            ).map_err(|error| error.to_string())?;
            Ok(report.status)
        }
        Mode::Determinism => {
            let report = run_determinism(
                &orchestration,
                &scenes,
                &DeterminismOptions {
                    repository_root,
                    output_dir: args.output,
                    battery_id: args.battery,
                    target: args.target.map(Into::into),
                },
            ).map_err(|error| error.to_string())?;
            Ok(report.status)
        }
    }
}
